### ISSUE TITLE
Spurious React key warnings if html macro chooses to interpret

### ISSUE THREAD

**Issue by dustingetz** [r0man/sablono#57]
In this diff, the red version generates React key warnings, and the green version does not.

![image](https://cloud.githubusercontent.com/assets/124158/6735465/b9c3e806-ce34-11e4-9c0a-d6ce075aecf5.png)

The `html` macro compile path does not generate warnings, but if the macro takes the interpreter path, it passes react an array rather than applying children as varargs.

**Comment by jmatsushita** [comment]
Got the same problem with [sablono "0.3.5-SNAPSHOT"]

**Comment by quetz** [comment]
Got same problem with sablono 0.5.3  So it seems that contents of every `let` block must be wrapped with `(html ...)` and it's pretty inconvenient when writing [rum](https://github.com/tonsky/rum) components. Any progress on this issue?

**Comment by r0man** [comment]
I added a test case for this issue over here https://github.com/r0man/sablono/pull/96, but can't reproduce the warning. Can someone provide a test case that does cause the warning?

**Comment by quetz** [comment]
Is it possible to get cljs repl within project? And also run one testcase (like `lein test :only ...`)

**Comment by quetz** [comment]
I'm complete clojure noob and was not able to assemble a test in sablono test suite, but here's minimal working example based on https://github.com/swannodette/mies template: https://github.com/quetz/sab57

To compile:
```
scripts/build
```
Then open `index.html`

**Comment by r0man** [comment]
@quetz You can get a REPL by running `lein figwheel`. Devcards are then also available at `http://localhost:3449`

**Comment by currentoor** [comment]
I'm also getting this error. Does anyone have a workaround?

**Comment by r0man** [comment]
@tonsky I'm trying to figure out what the problem is. Any idea why I can't `macroexpand-1` this:

``` clojure
(require '[sablono.core :refer [html]])
(require '[rum.core :refer [defc]])

(macroexpand-1
 '(defc issue-57-rum [text]
    (html
     (let [text-add (str text " warning")]
       [:div
        [:h1 text]
        [:h1 text-add]]))))
```

I get the following error from `parse-defc`.

```
1. Unhandled java.lang.IllegalArgumentException
   Syntax error at ([text] (html (let [text-add (str text " warning")]
   [:div [:h1 text] [:h1 text-add]])))

                      REPL:   25  rum.core/parse-defc
                      REPL:    9  rum.core/parse-defc
                      REPL:   31  rum.core/-defc
                      REPL:   30  rum.core/-defc
                  core.clj:   54  rum.core/defc
                  core.clj:   41  rum.core/defc
               RestFn.java:  490  clojure.lang.RestFn/invoke
                  Var.java:  401  clojure.lang.Var/invoke
                  AFn.java:  171  clojure.lang.AFn/applyToHelper
                  Var.java:  700  clojure.lang.Var/applyTo
             Compiler.java: 6776  clojure.lang.Compiler/macroexpand1
                  core.clj: 3894  clojure.core/macroexpand-1
                  core.clj: 3888  clojure.core/macroexpand-1
                      REPL:    7  sablono.expand/eval29655
                      REPL:    7  sablono.expand/eval29655
             Compiler.java: 6927  clojure.lang.Compiler/eval
             Compiler.java: 6890  clojure.lang.Compiler/eval
                  core.clj: 3105  clojure.core/eval
                  core.clj: 3101  clojure.core/eval
                  main.clj:  240  clojure.main/repl/read-eval-print/fn
                  main.clj:  240  clojure.main/repl/read-eval-print
                  main.clj:  258  clojure.main/repl/fn
                  main.clj:  258  clojure.main/repl
                  main.clj:  174  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   87  clojure.tools.nrepl.middleware.interruptible-eval/evaluate/fn
                  AFn.java:  152  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  646  clojure.core/apply
                  core.clj: 1881  clojure.core/with-bindings*
                  core.clj: 1881  clojure.core/with-bindings*
               RestFn.java:  425  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   85  clojure.tools.nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   55  clojure.tools.nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  222  clojure.tools.nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
    interruptible_eval.clj:  190  clojure.tools.nrepl.middleware.interruptible-eval/run-next/fn
                  AFn.java:   22  clojure.lang.AFn/run
   ThreadPoolExecutor.java: 1142  java.util.concurrent.ThreadPoolExecutor/runWorker
   ThreadPoolExecutor.java:  617  java.util.concurrent.ThreadPoolExecutor$Worker/run
               Thread.java:  745  java.lang.Thread/run
```

**Comment by tonsky** [comment]
No need to do do html inside sefc -- it's added by default by rum. Maybe that's the problem 

-----Original Message-----

**Comment by r0man** [comment]
@quetz I'm not sure but I think this is already fixed in master. Can you please try your example with sablono `0.6.0`. I don't get the warning with the latest version.

**Comment by quetz** [comment]
@tonsky, nope, removing `html` call would always result in warnings if there's a `let` block inside component. For example:
```
(rum/defc comp-ok [text]
  (let [text-add (str text " OK")]
    [:div
       [:h1 text]
       [:h1 text-add]
       ]))
```
Gives warning.

And this does not:
```
(rum/defc comp-ok [text]
  [:div
    [:h1 text]
    [:h1 text]
  ])
```

**Comment by quetz** [comment]
@r0man, yeah it seems to be fixed in 0.6.0. Thanks alot!

**Comment by r0man** [comment]
@quetz Ok, cool :)
